### PR TITLE
Expose details for basictracer spans / tracers

### PR DESCRIPTION
**Pull request by jmacd**
This allows basictracer implementations to extract detail (Context, Operation, Start) from spans at runtime and to retrieve the global Recorder implementation.

### CODE REVIEW

File: raw.go [hunk 1]
```diff
@@ -31,3 +31,32 @@ type RawSpan struct {
...
+
+	// Start indicates when the span began.
+	Start time.Time
+}
+
+// GetSpanDetail returns the SpanDetail and true if `span` is a
+// basictracer span, or false if `span` is not a basictracer
+// span.
+func GetSpanDetail(span opentracing.Span) (SpanDetail, bool) {
+	bspan, is := span.(*spanImpl)
```
Comment (bg451) [inline raw.go]: I'd prefer `s/is/ok/g`, since it's easier to read.

File: raw.go [hunk 2]
```diff
@@ -31,3 +31,32 @@ type RawSpan struct {
 	// The span's associated baggage.
 	Baggage map[string]string // initialized on first use
 }
+
+// SpanDetail summarizes the identifying fields of a Span, for
+// diagnostic and implementation-specific purposes.
+type SpanDetail struct {
```
Comment (yurishkuro) [inline raw.go]: why not define this as an interface and let basictracer's Span implement it? Then you don't need to copy anything, just convert. And the `GetSpanDetail` becomes unnecessary, users can simply do a checked cast to this new interface.

File: raw.go (hunk 1 above)
Comment (jmacd) [inline raw.go]: Done

File: tracer.go [hunk 3]
```diff
@@ -210,3 +210,16 @@ func (t *tracerImpl) Join(operationName string, format interface{}, carrier inte
 	}
 	return nil, opentracing.ErrUnsupportedFormat
 }
+
+// GetGlobalTracerOptions attempts to get the basictracer.Options
+// struct provided to the global Tracer.  Returns the options and true
+// if the global Tracer is a basictracer implementation; returns
+// false otherwise.
+func GetGlobalTracerOptions() (Options, bool) {
```
Comment (yurishkuro) [inline tracer.go]: why is this needed?

File: raw.go (hunk 2 above)
Comment (bg451) [inline raw.go]: Currently, the `Raw` field is unexported in `spanImpl`, so if anything we can make that exported and not need any of this code.

Comment (yurishkuro) [inline raw.go]: if `spanImpl` implements an interface, it won't matter
```go
func (s *spanImpl) GetContext() Context {
    return s.raw.Context
}
```

Comment (jmacd) [inline raw.go]: This approach: N+K fields are copied, there is one new struct definition, one new method definition, one type assertion, and one method call.
That approach: K fields are copied once, there is one new interface definition, N new method definitions, one type assertion, and K method calls.

(where K (number of fields accessed) is <= N (number of detail fields))

I'm not sure there is a significant difference either way, but it feels more straightforward to me this way. In our use-case, we have N=3 (Context, Operation, Start) and K=2 (Context, Start), fwiw.

File: tracer.go (hunk 3 above)
Comment (jmacd) [inline tracer.go]: Ultimately, we are trying to get from the result of opentracing.GetGlobalTracer() to the implementation of basictracer.Recorder. Exposing the Options seems useful and gives me a pointer to the Recorder. What would you suggest?

Following your earlier comment, I could make an interface to obtain the Options or the Recoder from a basictracer-supplied Tracer.

File: raw.go (hunk 2 above)
Comment (jmacd) [inline raw.go]: @bg451 Raw could be exported, but it would export a bunch of fields that are not safe to use (e.g., the maps) or not meaningful (e.g., duration).

Comment (jmacd) [inline raw.go]: @yurishkuro Sure. I'm willing to do it that way. But it copies a Context either way. I was taught to be suspicious of interfaces when there is only one implementation.

File: tracer.go (hunk 3 above)
Comment (yurishkuro) [inline tracer.go]: but why do you need access to Recorder? Regular application code should not need it. If you need it in tests, then presumably those same tests are actually in charge of creating the instance of the Recorded in the first place, so they could just keep that ref.

File: raw.go (hunk 2 above)
Comment (yurishkuro) [inline raw.go]: well, you seem to have a conflict of goals, you want to keep the fields private, yet you need to access them for some reason. Interface provides a projection, controlled access. But worth asking why there's a conflict in the first place :-)

File: tracer.go (hunk 3 above)
Comment (jmacd) [inline tracer.go]: This is not for regular application code. This is for our implementation to access some of its internal state (an API key) given a Tracer handle, to use for a special-purpose API. Since the Tracer is actually implemented by basictracer, we have no direct access to its state.

It's also not for test code. We know the internal state at the time we construct the Tracer. Getting from Tracer to the internal state is what we're trying to do.

File: raw.go (hunk 2 above)
Comment (jmacd) [inline raw.go]: (Suddenly realizing it would be better if my first pull request to an OT repo were not trying to punch holes in these abstractions.)  I would ask @bensigelman to fill in some motivation here, since I'm not sure I'll get the language right.

Comment (bensigelman) [inline raw.go]: I could weigh in on this debate, but I have another idea entirely... how about creating an interface like so:

```
type BasicSpan {
    opentracing.Span

    OperationName() string
    Start() time.Time
    Context() Context
}
```

... then the ultimate user of the basictracer-derived span would do something like

```
basicSpan := otSpan.(basictracer.BasicSpan)
fmt.Println("I can read whatever I want!", basicSpan.Operation())
...
```

Just to make sure I wasn't misremembering how Go works, I wrote up this example: http://play.golang.org/p/1KllYjHi2_

File: raw.go [hunk 4]
```diff
@@ -31,3 +31,32 @@ type RawSpan struct {
...
+	Operation string
+
+	// Start indicates when the span began.
+	Start time.Time
+}
+
+// GetSpanDetail returns the SpanDetail and true if `span` is a
+// basictracer span, or false if `span` is not a basictracer
+// span.
+func GetSpanDetail(span opentracing.Span) (SpanDetail, bool) {
```
Comment (bensigelman) [inline raw.go]: given my proposal above, the `bool` return value would just be the second return value in the cast:

```
basicSpan, ok := otSpan.(basictracer.BasicSpan)
```

`</obvious>`

File: tracer.go (hunk 3 above)
Comment (bensigelman) [inline tracer.go]: @yurishkuro In our case, for instance, we want to allow the programmer/user to interact with our actual *tracing system* from their application code (not just describe the instrumentation)... e.g., request/trigger trace assembly for an arbitrary Span available within the process.

FWIW, we don't *need* to be able to access the `Recorder` (and thus our basictracer impl, basically)... we could ask the programmer to keep their own reference to our tracing system on hand. But it would be more convenient if that user only had to pass around a single (Tracer) object.

Also, just to clarify: none of the casting here affects the trace structure per se... it's all about getting access to things like span guids/etc, then using them to trigger effects in other systems. (If we were going through this trouble just to modify the trace structure itself I would consider this sort of downcasting very non-OT-ish)

Okay, on balance: I think there is good reason to expose some sort of read API on basictracer Spans. Though we could try leaving the Recorder/Options out of this change and see how far we can get by asking our own users to maintain a separate reference to our tracing system (independent of the `Tracer` which is just used to add instrumentation).

File: raw.go (hunk 2 above)
Comment (bg451) [inline raw.go]: @yurishkuro Looking back at my previous comment, I just realized that I forgot that `spanImpl` is unexported, so you can't cast it outside of the package anyways. 

I'm +1 on the interface approach, since it'd be nice to avoid adding extra exported structs to the package.

File: tracer.go (hunk 3 above)
Comment (bg451) [inline tracer.go]: I just took a look at the PR in the lightstep-tracer-go repo and have a much better idea what this PR is trying to accomplish. 

@jmacd How about generalizing this, so instead it's `GetTracerOptions(tracer opentracing.Tracer)...`

File: raw.go (hunk 2 above)
Comment (bensigelman) [inline raw.go]: PS: my idea is just what Yuri suggested... I misread his comment the first time around.

**Comment by jmacd** [comment]
Hey all. I apologize for sending this PR yesterday when it was clearly not well reviewed on our end. I misinterpreted the directions I was given, but now that I'm here we might as well continue (and sorry for all that I'm clearly still learning the workflow).

We need at least a way to get the SpanID from a Span, and ideally a way to refer to the Tracer implementation. We discussed the feedback above and settled on the latest snapshot. I have some reservations, personally, about using interfaces like this, as the methods are to gain access to implementation details for use by an implementation--concealing them behind interfaces feels like premature generalization. I'd rather see us export the concrete basictracer Span/Tracer types and some (not all) of their fields.

File: raw.go (hunk 2 above)
Comment (bensigelman) [inline raw.go]: @jmacd made the good point this morning that – since the impl objects are immediately recycled by the pool – any access provided via a type conversion to something akin to the `BasicSpan` interface above would have to come along with strong caveats about calling any methods prior to `Finish()`.

Anyway, I still like the interface approach.

File: span.go [hunk 5]
```diff
@@ -9,6 +9,25 @@ import (
...
 )
 
+// Span provides access to the essential details of the span, for use
+// by basictracer consumers.  These methods may only be called prior
+// to (*opentracing.Span).Finish().
+type Span interface {
+	opentracing.Span
+
+	// TraceID is a globally unique trace identifier
+	TraceID() int64
```
Comment (bensigelman) [inline span.go]: Inasmuch as a `basictracer.Context` concept exists, I think I'd prefer that we just expose it here via `Context() Context`, copy bedamned. This is a nod to the fact that ("many moons ago") `Context` was pluggable and probably should be again in the future... i.e., initializing a `basictracer.Tracer` would require not just a `Recorder` but also some sort of context provider + context codec.

My two cents... (Though if we're going to expose traceid / spanid directly, I support your decision to leave the parentid and sampled status out of it for now)

**Comment by bensigelman** [comment]
@jmacd I made one comment... curious what others think. I guess it's a question about the intended scope of `basictracer`... if we want it to be opinionated about what constitutes a `Context`, then the change you have here LGTM. If we want to hedge our bets and be more general, then there's an argument to make the API more general. I could be swayed either way but wanted to raise the topic.

**Comment by jmacd** [comment]
@bensigelman You mentioned the idea of a context provider and codec. I feel that if this is an option on the table, it will influence the shape of this API and make much of the above discussion meaningless. E.g., will the provided Context have its lifetime tied to the Span? I would guess not. There will be less incentive to copy and more incentive to take a reference. Then we might have something like

```
rspan := span.(basictracer.Span)
ctx := rspan.Context()
span.Finish()

var id [256]byte
if !ctx.GetID(buf) {
  return errNoSpanID
}
Pseudocode.doSomethingWith(string(id))
```

File: tracer.go [hunk 6]
```diff
@@ -6,6 +6,15 @@ import (
 	opentracing "github.com/opentracing/opentracing-go"
 )
 
+// Tracer extends the opentracing.Tracer interface with methods to
+// probe implementation state, for use by basictracer consumers.
+type Tracer interface {
+	opentracing.Tracer
+
+	// GetOptions gets the Options used in New() or NewWithOptions().
+	GetOptions() Options
```
Comment (bensigelman) [inline tracer.go]: nit: Go accessors typically don't have the `Get` prefix, right?

Comment (bg451) [inline tracer.go]: Yeah that's right.

**Comment by bensigelman** [comment]
@jmacd I am ok with what you have here (LOKTM? :), so fine to merge unless @yurishkuro/@bg451/@tschottdorf have objections in the next 24h.

I would still slightly prefer `basictracer.Span.Context()` over the per-field methods (per [my comment from a few days back](https://github.com/opentracing/basictracer-go/pull/18/files#r57405047)), but I can see arguments either way and don't feel strongly.

Re the future of a Context provider/codec: what you're suggesting makes sense, but I'd like to get more experience with basictracer in production before making more substantial changes to the API. (Happy to discuss further if that doesn't make sense)

**Comment by bg451** [comment]
lgtm

File: tracer.go (hunk 6 above)
Comment (jmacd) [inline tracer.go]: The type already has a (embedded) field named Options. I'll rename it "options"

**Comment by jmacd** [comment]
@bhs OK, I added Context() in place of SpanID() and TraceID() since (you asked and) that's more compatible w/ future abstraction.

**Comment by bensigelman** [comment]
thanks!

**Comment by bg451** [comment]
@jmacd I forgot to add this should be squashed.

**Comment by jmacd** [comment]
@bg451 (I'm confused) What should be squashed?

**Comment by bg451** [comment]
All of the commits before merging. It keeps the master commit log nice and
tidy. `git rebase -i`

**Comment by jmacd** [comment]
Done
